invenio_rdm_records/services/customizations.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Helpers for customizing the configuration in a controlled manner."""
+
+
+#
+# Helpers
+#
+def _make_cls(cls, attrs):
+    """Make the custom config class."""
+    return type(f'Custom{cls.__name__}', (cls, ), attrs, )
+
+
+#
+# Mixins
+#
+class SearchOptionsMixin:
+    """Customization of search options."""
+
+    @classmethod
+    def customize(cls, opts):
+        """Customize the search options."""
+        attrs = {}
+        if opts.facets:
+            attrs['facets'] = opts.facets
+        if opts.sort_options:
+            attrs['sort_options'] = opts.sort_options
+            attrs['sort_default'] = opts.sort_default
+            attrs['sort_default_no_query'] = opts.sort_default_no_query
+        return _make_cls(cls, attrs) if attrs else cls
+
+
+class FileConfigMixin:
+    """Shared customization for file service configs."""
+
+    @classmethod
+    def customize(cls, permission_policy=None):
+        """Class method to customize the config for an instance."""
+        attrs = {}
+        # Permission policy
+        if permission_policy:
+            attrs['permission_policy_cls'] = permission_policy
+        # Create the config class
+        return _make_cls(cls, attrs) if attrs else cls
+
+
+class RecordConfigMixin:
+    """Shared customization for record service configs."""
+
+    @classmethod
+    def customize(cls, permission_policy=None, doi_registration=False,
+                  **kwargs):
+        """Class method to customize the config for an instance."""
+        attrs = {}
+
+        # Permission policy
+        if permission_policy:
+            attrs['permission_policy_cls'] = permission_policy
+
+        # Search options
+        for opt in ['search', 'search_drafts', 'search_versions']:
+            if opt in kwargs:
+                search_opt_cls = getattr(cls, opt)
+                attrs[opt] = search_opt_cls.customize(kwargs[opt])
+
+        # Create the config class
+        new_cls = _make_cls(cls, attrs)
+
+        # TODO: Change how to customize PID providers. Was moved here from
+        # ext.py
+        if not doi_registration:
+            new_cls.pids_providers.pop("doi", None)
+
+        return new_cls
```

Comment:
Ouf... This is to achieve: `RDM<Service>Config` -> set some attributes -> use this configuration with the set attributes but leave the global `RDM<Service>Config` untouched so not mixing with others, right?

What about just instantiating the config class? 

```python
class CustomizableConfigMixin:
    """Mixin to allow local config modifications."""

    @classmethod
    def customize(cls, customizations**):
        """Class method to customize the config for an instance."""
        config = cls()
        for attribute, value in customizations.items():
            if attribute in config:  # if we only want to allow customizations of existing attributes
                setattr(config, attribute, value)
        return config
```  

The instance of the config class has all the same attributes, but now their modification is local to that instance. `customize` can also be edited to restrict the fields that can be customized if so desired, although leaving it open would make any future additions easier.

When we drop support for Python 3.6, we can even use [dataclasses](https://docs.python.org/3.7/library/dataclasses.html).

Reply:
> Ouf... This is to achieve: RDM<Service>Config -> set some attributes -> use this configuration with the set attributes but leave the global RDM<Service>Config untouched so not mixing with others, right?

Yes, the intention is to leave it untouched.

1. I agree that instantiation of the config class ``RDMRecordsService(permission_policy, ...)`` is a better way to customize the class. The problem is I find it very dirty if we start sometimes passing classes and other times we pass instances, so I'd prefer to keep it like this and add a comment in the code that ideally this should refactor just be a class instantiation.

2. It's intentional that you can only customize some specific attributes (to keep a clean API of which customizations are supported). I.e. I want to make it a conscious decision that something is customizable.